File: python/pixzig/paths.py
"""Asset path resolution for the Python bindings.

Every relative path handed to a loader -- textures, tilemaps, fonts, sounds,
animation files, the manifest -- is resolved against **the directory holding
the main script**, not the process's current working directory. That way a
game runs the same whichever directory it was launched from:

    python examples/hello_pixzig.py        # from the repo root
    cd examples && python hello_pixzig.py  # same assets either way

The Zig side resolves its own relative paths against the executable's
directory (`pixzig.paths`), which under Python is wherever the interpreter
happens to be installed. So the bindings resolve to an absolute path before
crossing the FFI; absolute paths pass through both layers untouched.

Pass `asset_root=` to `App` (or assign `app.asset_root`) to point somewhere
else -- a packaged install, a mod directory, a path read from a config file.
"""
import os
import sys


def main_script_dir() -> str:
    """The directory holding the `__main__` script, or the cwd when there
    isn't one (an interactive session, `python -c`, some frozen builds)."""
    main = sys.modules.get("__main__")
    script = getattr(main, "__file__", None)
    if script:
        return os.path.dirname(os.path.abspath(script))
    return os.getcwd()


class AssetPaths:
    """Resolves relative asset paths against `root`.

    One of these is created per `App`; the subsystems that load files
    (`Audio`, `Text`, and the loaders on `App` itself) share it, so changing
    `app.asset_root` affects every later load.
    """

    def __init__(self, root: str = None):
        self.root = root if root is not None else main_script_dir()

    @property
    def root(self) -> str:
        """The base directory, always absolute; assigning a relative path
        absolutizes it against the cwd at that moment."""
        return self._root

    @root.setter
    def root(self, value: str) -> None:
        self._root = os.path.abspath(value)

    def resolve(self, path: str) -> str:
        """Returns `path` as an absolute path, relative paths taken from
        `root`. Already-absolute paths are returned unchanged."""
        if os.path.isabs(path):
            return path
        return os.path.normpath(os.path.join(self.root, path))

    def encode(self, path: str) -> bytes:
        """`resolve`, UTF-8 encoded for the FFI."""
        return self.resolve(path).encode("utf-8")

File: src/pixzig/paths.zig
//! Asset path resolution.
//!
//! Every relative path a game hands the engine -- textures, atlases, fonts,
//! tilemaps, sounds, scripts, manifests -- is resolved against the directory
//! the executable lives in (`SDL_GetBasePath`), not the process's current
//! working directory. That way a packaged game finds its assets no matter
//! which directory it was launched from, and `cd` in a shell never changes
//! which files a build loads.
//!
//! Absolute paths are passed through untouched, so a game that computes its
//! own paths (a mod directory, a save folder, a path from a config file)
//! keeps full control.

const std = @import("std");
const platform = @import("./platform.zig");

/// The build's asset base directory, or null to use the executable's own
/// directory. `buildEngine`/`buildGame` generate this module: a dev build
/// points it at the source tree so hot-reload watches the real files, a
/// packaged build leaves it null.
const asset_base: ?[]const u8 = @import("pixzig_asset_base").dir;

/// The directory relative asset paths resolve against: the build's
/// `pixzig_asset_base` when it set one, otherwise the directory holding the
/// running executable (`SDL_GetBasePath`, with a trailing path separator,
/// borrowed from SDL for the life of the process).
///
/// Null under Emscripten, where the virtual filesystem is rooted at `/` and
/// a relative path is already correct as written.
pub fn baseDir() ?[]const u8 {
    return asset_base orelse platform.basePath();
}

/// Resolves `path` to the file the engine should actually open. Caller owns
/// the returned buffer.
pub fn resolve(alloc: std.mem.Allocator, path: []const u8) ![]u8 {
    if (std.fs.path.isAbsolute(path)) return alloc.dupe(u8, path);
    const base = baseDir() orelse return alloc.dupe(u8, path);
    return std.fs.path.join(alloc, &.{ base, path });
}

/// `resolve`, null-terminated for the C libraries the engine hands paths to
/// (stbi, miniaudio, Lua). Caller owns the returned buffer.
pub fn resolveZ(alloc: std.mem.Allocator, path: []const u8) ![:0]u8 {
    if (std.fs.path.isAbsolute(path)) {
        return std.mem.concatWithSentinel(alloc, u8, &.{path}, 0);
    }
    const base = baseDir() orelse return std.mem.concatWithSentinel(alloc, u8, &.{path}, 0);
    return std.fs.path.joinZ(alloc, &.{ base, path });
}
